### PR TITLE
Introduce the Flower Enhancement Docs process

### PR DESCRIPTION
**Pull request by nfnt**
#### What does this implement/fix? Explain your changes.

To document motivations and decisions for larger changes or features to Flower, an Enhancement Document process (FED) is proposed. This process is heavily influenced by similar processes in the Python language (https://peps.python.org) as well as in Kubernetes (https://git.k8s.io/enhancements/keps).
The goal of this process is to provide insight and transparency on upcoming features that might change how users or developers have to interact with Flower. By storing proposals as part of Flower's repository, the community has an easy and accessible way to understand and influence these important changes to Flower.

#### Any other comments?

This is very similar in style to the [Kubernetes Enhancement Proposal Process](https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/0000-kep-process/README.md). That process has been slightly simplified, as some parts only made sense in a project as large as Kubernetes.

### CODE REVIEW

File: proposals/0001-20220311-flower-enhancement-proposals.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+### Goals
+
+Roughly any larger, user facing enhancement should follow the Enhancement Proposal process. If an enhancement would be described in either written or verbal communication to anyone besides the author or developer, then consider creating an Enhancement Proposal.
+
+Similarly, any technical effort (refactoring, major architectural change) that will impact a large section of the development community should also be communicated widely. The Enhancement Proposal process is suited for this even if it will have zero impact on the typical user or operator.
+
+### Non-Goals
+
+For small changes and additions, going through the Enhancement Proposal process would be time-consuming and unnecessary. This, e.g., includes adding new Federated Learning algorithms, as these only add features without changing how flower works or is used.
```
Comment (danieljanes) [inline proposals/0001-20220311-flower-enhancement-proposals.md:58]:
```suggestion
For small changes and additions, going through the Enhancement Proposal process would be time-consuming and unnecessary. This includes, for example, adding new Federated Learning algorithms, as these only add features without changing how Flower works or is used.
```

File: proposals/0001-20220311-flower-enhancement-proposals.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+- __status__ (Required)  
+  The current status of the proposal. See [workflow](#workflow) for the possible states.
+- __authors__ (Required)  
+  A list of authors of the proposal. This is simply the Github ID.
+- __creation-date__ (Required)  
+  The date that the proposal was first submitted in a PR.
+- __last-updated__ (Optional)
+  The date that the proposal was last changed significantly.  
+- __see-also__ (Optional)  
+  A list of other proposals that are releavent to this one.
```
Comment (danieljanes) [inline proposals/0001-20220311-flower-enhancement-proposals.md:102]:
```suggestion
  A list of other proposals that are relevant to this one.
```

File: proposals/0001-20220311-flower-enhancement-proposals.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+- __replaces__ (Optional)  
+  A list of proposals that this one replaces.
+- __superseded-by__ (Optional)  
+  A list of proposals that this one supersedes.
+
+### Workflow
+
+The idea forming the enhancement should already have been discussed or pitched in the comminuty. As such, it needs a champion, usually the author, who shepherds the proposal. This person also has to find committers to flower willing to review the proposal.
```
Comment (danieljanes) [inline proposals/0001-20220311-flower-enhancement-proposals.md:110]:
```suggestion
The idea forming the enhancement should already have been discussed or pitched in the comminuty. As such, it needs a champion, usually the author, who shepherds the proposal. This person also has to find committers to Flower willing to review the proposal.
```

File: proposals/0001-20220311-flower-enhancement-proposals.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
+---
+ep-number: 0001
+title: Flower Enhancement Proposals
```
Comment (danieljanes) [inline proposals/0001-20220311-flower-enhancement-proposals.md:3]: Flower Enhancements Design?
Flower Enhancements Doc?

Comment (orlandohohmeier) [reply proposals/0001-20220311-flower-enhancement-proposals.md:3]: I like "Flower Enhancement Doc"

**Comment by orlandohohmeier** [comment]
@danieljanes addressed your comments and adjust the name as suggested. Please have another look.

File: enhancements/0001-20220311-flower-enhancement-doc.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Enhancement Proposal Template](#enhancement-proposal-template)
+  - [Metadata](#metadata)
+  - [Workflow](#workflow)
+- [Drawbacks](#drawbacks)
+- [Alternatives Considered](#alternatives-considered)
+  - [Github Issues](#github-issues)
```
Comment (danieljanes) [inline enhancements/0001-20220311-flower-enhancement-doc.md:25]:
```suggestion
  - [GitHub Issues](#github-issues)
```

File: enhancements/0001-20220311-flower-enhancement-doc.md [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+- a feature, and effort-tracking document
+- a product requirements document
+- a design document
+
+into one file, which is created incrementally in collaboration with the community.
+
+## Motivation
+
+For far-fetching changes or features proposed to Flower, an abstraction beyond a single Github issue or pull request is required in order to understand and communicate upcoming changes to the project.
```
Comment (danieljanes) [inline enhancements/0001-20220311-flower-enhancement-doc.md:49]:
```suggestion
For far-fetching changes or features proposed to Flower, an abstraction beyond a single GitHub issue or pull request is required in order to understand and communicate upcoming changes to the project.
```

File: enhancements/0001-20220311-flower-enhancement-doc.md [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+### Metadata
+
+- **fed-number** (Required)
+  The `fed-number` of the last Flower Enhancement Doc + 1. With this number, it becomes easy to reference other proposals.
+- **title** (Required)
+  The title of the proposal in plain language.
+- **status** (Required)
+  The current status of the proposal. See [workflow](#workflow) for the possible states.
+- **authors** (Required)
+  A list of authors of the proposal. This is simply the Github ID.
```
Comment (danieljanes) [inline enhancements/0001-20220311-flower-enhancement-doc.md:99]:
```suggestion
  A list of authors of the proposal. This is simply the GitHub ID.
```

File: enhancements/0001-20220311-flower-enhancement-doc.md [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+- a product requirements document
+- a design document
+
+into one file, which is created incrementally in collaboration with the community.
+
+## Motivation
+
+For far-fetching changes or features proposed to Flower, an abstraction beyond a single Github issue or pull request is required in order to understand and communicate upcoming changes to the project.
+
+The purpose of this process is to reduce the amount of "tribal knowledge" in our community. By moving decisions from Slack threads, video calls and hallway conversations into a well tracked artifact, this process aims to enhance communication and discoverability.
```
Comment (danieljanes) [inline enhancements/0001-20220311-flower-enhancement-doc.md:51]:
```suggestion
The purpose of this process is to reduce the amount of "tribal knowledge" in our community. By moving decisions from Slack threads, video calls, and hallway conversations into a well tracked artifact, this process aims to enhance communication and discoverability.
```

**Review by danieljanes — CHANGES_REQUESTED** [review]

**Review by tanertopal — APPROVED** [review]

**Review by danieljanes — APPROVED** [review]

**Comment by danieljanes** [comment]
@nfnt auto-merge is enabled, it'll automatically be merged once you update the branch with the latest changed from `main`

**Review by danieljanes — APPROVED** [review]

**Pull request merged**